### ISSUE TITLE
Trudesk out of memory ( aws t2.medium ec2 )

### ISSUE THREAD

**Issue by earteza** [polonel/trudesk#117]
<--- Last few GCs --->

[9305:0x3252d90]  3807303 ms: Scavenge 1371.1 (1423.1) -> 1371.0 (1424.1) MB, 14.4 / 0.0 ms  (average mu = 0.208, current mu = 0.139) allocation failure
[9305:0x3252d90]  3807325 ms: Scavenge 1371.5 (1424.1) -> 1371.3 (1424.6) MB, 5.5 / 0.0 ms  (average mu = 0.208, current mu = 0.139) allocation failure
[9305:0x3252d90]  3807371 ms: Scavenge 1371.9 (1424.6) -> 1371.5 (1425.6) MB, 14.9 / 0.0 ms  (average mu = 0.208, current mu = 0.139) allocation failure

<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x3bc7031dbe1d]
Security context: 0x3c532491e6e1 <JSObject>
    1: keys [0x3c5324905199](this=0x3c5324904631 <JSFunction Object (sfi = 0xd02418e301)>,0x05247d65c371 <Object map = 0x356c1b882571>)
    2: isExclusive(aka isExclusive) [0x202087b916c9] [/etc/trudesk/node_modules/mongoose/lib/helpers/projection/isExclusive.js:~9] [pc=0x3bc704c5a2a0](this=0x1b2719a026f1 <undefined>,projection=0x05247d65c371 <Object map = 0...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x8daaa0 node::Abort() [node]
 2: 0x8daaec  [node]
 3: 0xad73ce v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0xad7604 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0xec4c32  [node]
 6: 0xec4d38 v8::internal::Heap::CheckIneffectiveMarkCompact(unsigned long, double) [node]
 7: 0xed0e12 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [node]
 8: 0xed1744 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
 9: 0xed43b1 v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationSpace, v8::internal::AllocationAlignment) [node]
10: 0xe9d834 v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [node]
11: 0x113cf9e v8::internal::Runtime_AllocateInNewSpace(int, v8::internal::Object**, v8::internal::Isolate*) [node]
12: 0x3bc7031dbe1d

**Comment by polonel** [comment]
What version of Node?
What version of Trudesk?

**Comment by earteza** [comment]
> What version of Node?
> What version of Trudesk?

node: v10.13.0
trudesk: 1.0.2
CPU usage reach 100%

**Comment by polonel** [comment]
I'm looking into this as I have never encountered it. The demo data I use for all development has over 100k active tickets. 
How much memory do you have on the machine running trudesk?

My best guess right now is that its the `cache` process that is hitting 100% cpu. The main process spwns a fork to load a cache of the tickets over the last 365 days. 

If you enable `NODE_ENV=development` it will show debug output in the log. Match the PID of the `[Child:Cache:PID]` to your process list to confirm it is that fork (PID) that is hitting 100% CPU.

**Comment by polonel** [comment]
Also what OS and Version? 

I'm trying to reproduce.

**Comment by earteza** [comment]
@polonel 
node process:
```ps -p 10389 -o %cpu,%mem,cmd
%CPU %MEM CMD
16.4  4.4 node app.js```

**Comment by polonel** [comment]
Thank you for the information.

I have a couple of bots I created to help simulate multiple user usage. I'm looking into reproducing any type of CPU / Mem Leak.

**Comment by earteza** [comment]
@polonel  is it possible to share your bot with us, so that we can also check from here?

**Comment by polonel** [comment]
@earteza I have to need to modify them a little before I can run them. The bots can't run in production, as it creates tickets and deletes them, same for users, etc. 

I have a couple of questions to help me rule out a couple of things.

Do you utilize the built-in chat?
Does your tickets have attachments?
If the active users refresh their browser window, does the CPU usage drop?
If all users log out does the CPU stay at its current %?

I have a guess where the bug may be coming from but won't know for sure until I get the bots up and running.

**Comment by polonel** [comment]
@earteza I'm still investigating and not detecting anything.
Is trudesk running behind a proxy (Nginx)?
Is there a minimum number of users active that you start to see the issue?
I know you said it was over time, how long is that time?

**Comment by polonel** [comment]
@earteza 
I run ubuntu 16.04 in production. I have multiple clients running Ubuntu and some with around 10-15 active users. I checked with them today and they haven't experienced the issue. The only difference is they are all on MongoDB 3.6.

I'm setting up a deployment with MongoDB 4.0 to give it a test. The highest I've seen CPU is around 10% but only while processing cache, and the highest I've seen memory is around 250mb. The memory usage goes down after GC happens though. 

Going to test with MongoDB 4.0.

**Comment by earteza** [comment]
@polonel  

Can you send me your server information? 

Another question: is there is any balancing solution when we have around 200~250 users?

**Comment by polonel** [comment]
@earteza 
**Another question: is there is any balancing solution when we have around 200~250 users?**
My cloud platform is built off kubernetes where I've scaled trudesk to run with multiple containers and load balance between them. Even across regions. This has been in an isolated test environment which still needs some work with MongoDB sharding.

**Comment by earteza** [comment]
we use that script too, but still can't figure out what is the issue. so we keep testing on windows server see anything we get from there .

**Comment by polonel** [comment]
@earteza 
Yes, I'm determined to try and reproduce. It's possible to be an issue with the version of mongoose I use and MongoDB 4.0

I'm testing/researching.

**Comment by earteza** [comment]
we testing on a windows server, having the same issue there.

**Comment by polonel** [comment]
With MongoDB 4? 

What are your reproduce steps?
I still haven’t seen the error occur. 
Was this with your existing data? How big is your data? On average how many comments/notes does a ticket have?

**Comment by earteza** [comment]
@polonel 

on windows server:
we install latest mongoDB in website
there is no error or anything. CPU just raise up. my data is not big only 450 tickets few comments.

**Comment by polonel** [comment]
Have you or can you try with MongoDB 3.6?

**Comment by earteza** [comment]
@polonel oky,  wait
let's me reinstall

**Comment by polonel** [comment]
This is what I'm seeing with a production site with 1400 tickets with 8 active users on right now. I looked at the process every hour.

This is running in a Ubuntu 16.04. VM with 4gb RAM and MongoDB 3.6.9 installed on same server.

``` bash
root@ubuntu:~# ps -p 6782 -o %cpu,%mem,cmd
%CPU %MEM CMD
 2.5 13.7 node app

root@ubuntu:~# ps -p 6782 -o %cpu,%mem,cmd
%CPU %MEM CMD
 2.5 13.8 node app

root@ubuntu:~# ps -p 6782 -o %cpu,%mem,cmd
%CPU %MEM CMD
 2.6 18.0 node app
```

**Comment by polonel** [comment]
I'm currently at a loss. I'm guessing something with your data is corrupt and it's having trouble processing.

This is with MongoDB 4.0 with an uptime of 5, 6, 7 hours. Active 4 users during the timeframe. 

``` bash
root@ubuntu:~# ps -p 6782 -o %cpu,%mem,cmd,time
%CPU %MEM CMD                             TIME
 1.4 16.7 node app                    00:05:44

root@ubuntu:~# ps -p 6782 -o %cpu,%mem,cmd,time
%CPU %MEM CMD                             TIME
 1.4 14.9 node app                    00:06:45

root@ubuntu:~# ps -p 6782 -o %cpu,%mem,cmd,time
%CPU %MEM CMD                             TIME
 1.6 13.8 node app                    00:07:44
```

**Comment by earteza** [comment]
@polonel we are keep testing it will update you soon.

**Comment by earteza** [comment]
@polonel can you let me know what is this button for ? 
![image](https://user-images.githubusercontent.com/45551566/49797163-dbbfb400-fd68-11e8-8612-4f1b784275f1.png)

**Comment by polonel** [comment]
@earteza It is a switch to auto refresh the ticket grid when tickets are created/updated. There are some cases where you may want to turn off the refresh event if you are in the process of checking many checkboxes on the ticket grid (During refresh it would remove your checked checkboxes).

Any further updates on the Out-of-Memory error?
